### PR TITLE
WIP: Bean functions binding

### LINKED ISSUE #905: @Bind_ annotations clarification
I've been looking at implementing the `@BindLombok` annotation as discussed in #902 yesterday. Keeping in mind that public fields are now also bound in `@BindBean` (from #895), it might get a little confusing as to what is bound where.

I suggest having 3 distinct annotations:
- `@BindBean` : like it used to work, binds bean properties (`public int getSomething()`)
- `@BindFields`: binds public fields (`public final int something;`)
- `@BindFunctions`: bind functions with no parameters (`public int binding() { return something; }`) (this would also work for Lombok's fluent accessors)

Anyway, just an idea, I was nearly done with the `@BindLombok` when I realized this could be used for a much larger use case with exactly the same code so I figured I might as well suggest it before making a PR. Thoughts?

### PR DESCRIPTION
**Pull request by FredDeschenes**
Fixes #905 and #902.

This add the `@BindFunctions` annotation and moves the public field bindings from `@BindBean` to their own `@BindFields` annotation.

I'll update the ASCIIdoc if/when you guys are ok with the result.

### CODE REVIEW

**Comment by FredDeschenes** [comment]
Yeah I wasn't sure on the naming, it's mostly why I pushed this before I finished everything :P

I'll have some time later in the week to fix this and update the documentation, I'll update the PR then.

**Comment by FredDeschenes** [comment]
That should do it  hopefully!

TODO:
- Decide on switching bindBean to bindProperties
- Update ASCIIdoc

File: docs/src/adoc/index.adoc [hunk 1]
```diff
@@ -330,7 +330,7 @@ You can bind multiple arguments from properties of a Java Bean:
 
 [source,java]
 ----
-Contact contact = Contact();
+Contact contact = new Contact();
```
Comment (qualidafial) [inline docs/src/adoc/index.adoc]: Heh, whoops! Good catch.

File: docs/src/adoc/index.adoc [hunk 2]
```diff
@@ -352,12 +385,13 @@ Document document =
...
       .bindBean("f", folder)
-      .bindBean("d", document)
+      .bindMethods("f", folder)
+      .bindFields("d", document)
       .execute();
 ----
 
 [WARNING]
-Neither `bindMap()` nor `bindBean()` support binding of nested properties (e.g.
+Neither `bindMap()`, `bindMethods()`, `bindFields()` nor `bindBean()` support binding of nested properties (e.g.
```
Comment (qualidafial) [inline docs/src/adoc/index.adoc]: Needs an Oxford comma between ``` `bindFields()` ``` and `nor`.

Also.. I know this is super pedantic, but can you line wrap this to follow 80-character line limit? I've been working carefully to keep the documentation format consistent.

Please read through the "Style guidelines" comment block at the top of this file if you haven't already, so you're aware of them.

File: docs/src/adoc/index.adoc [hunk 3]
```diff
@@ -1537,25 +1571,42 @@ In SQL Object (but not in Core), you can qualify a bound map with a prefix:
...
+
+Or you can bind from public, parameterless methods of an object:
+
+[source,java]
+----
+@SqlUpdate("insert into users (id, name) values (:functionThatReturnsTheId, :functionThatReturnsTheName)")
+void insert(@BindMethods User user);
+----
+
+Like `@BindMap`, `@BindBean` `@BindMethods` and `@BindFields` can have an optional prefix:
```
Comment (qualidafial) [inline docs/src/adoc/index.adoc]: Need to rephrase a little and add commas to be grammatically correct. Perhaps:

```asciidoc
Like `@BindMap`, the `@BindBean`, `@BindMethods`, and `@BindFields` annotations can
specify an optional prefix:
```

File: docs/src/adoc/index.adoc [hunk 4]
```diff
@@ -1537,25 +1571,42 @@ In SQL Object (but not in Core), you can qualify a bound map with a prefix:
...
 ----
 @SqlUpdate("insert into users (id, name) values (:user.id, :user.name)")
 void insert(@BindBean("user") User user);
+//void insert(@BindFields("user") User user);
+//void insert(@BindMethods("user") User user);
 ----
 
 [WARNING]
-As in the Core API, neither `@BindMap` nor `@BindBean` support binding of
+As in the Core API, neither `@BindMap`, `@BindBean`, `@BindFields` nor `@BindMethods` support binding of
```
Comment (qualidafial) [inline docs/src/adoc/index.adoc]: Let's add an Oxford comma between ``` `@BindFields` ``` and `nor`

File: docs/src/adoc/index.adoc [hunk 5]
```diff
@@ -2474,7 +2525,7 @@ interface, as its name suggests, finds arguments by name from some source.
 Typically a single `NamedArgumentFinder` instance will provide arguments for
 several different names.
 
-In cases where neither `bindBean()` nor `bindMap()` are a good fit, you can
+In cases where neither `bindBean()`, `bindFields()`, `bindMethods()` nor `bindMap()` are a good fit, you can
```
Comment (qualidafial) [inline docs/src/adoc/index.adoc]: Oxford comma here too please

File: docs/src/adoc/index.adoc [hunk 6]
```diff
@@ -2490,11 +2541,13 @@ stmt.bindNamedArgumentFinder(cacheFinder);
 
 [TIP]
 Under the hood, the
-link:{jdbidocs}/core/statement/SqlStatement.html#bindBean-java.lang.Object-[SqlStatement.bindBean()^]
+link:{jdbidocs}/core/statement/SqlStatement.html#bindBean-java.lang.Object-[SqlStatement.bindBean()^],
+link:{jdbidocs}/core/statement/SqlStatement.html#bindMethods-java.lang.Object-[SqlStatement.bindMethods()^],
+link:{jdbidocs}/core/statement/SqlStatement.html#bindFields-java.lang.Object-[SqlStatement.bindFields()^]
```
Comment (qualidafial) [inline docs/src/adoc/index.adoc]: And here please

File: docs/src/adoc/index.adoc [hunk 7]
```diff
@@ -2490,11 +2541,13 @@ stmt.bindNamedArgumentFinder(cacheFinder);
...
 and
 link:{jdbidocs}/core/statement/SqlStatement.html#bindMap-java.util.Map-[SqlStatement.bindMap()^]
 methods are just creating and binding custom implementations of
-`NamedArgumentFinder` for beans and maps, respectively.
+`NamedArgumentFinder` for beans, methods, fields and maps, respectively.
```
Comment (qualidafial) [inline docs/src/adoc/index.adoc]: And here. :)

**Comment by qualidafial** [comment]
Rebase or squash not necessary. We can squash merge the PR through the Github interface, but rebasing / squashing on your end makes it difficult to review code changes--no way to tell what are the new changes from the rest.

**Comment by FredDeschenes** [comment]
I think I got too used to working on my own repositories and rebasing as I please :P

Also thanks for making me realize my max-width line display thingy in IntelliJ was set to 120 chars and not 80!

**Comment by qualidafial** [comment]
Perfect. Thanks for all your work!

**Pull request merged**